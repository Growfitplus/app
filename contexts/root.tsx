import React from 'react';

interface ProvidersProps {
  providers: Array<React.JSXElementConstructor<React.PropsWithChildren<any>>>;
  children: React.ReactNode;
}

const RootProvider: React.FC<ProvidersProps> = ({ providers, children }) =>
  providers.reduceRight((next, Provider) => <Provider>{next}</Provider>, children);

export default RootProvider;
