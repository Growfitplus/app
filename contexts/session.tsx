import { useStorageState } from '@/hooks/useStorageState';
import { router } from 'expo-router';
import { useContext, createContext, type PropsWithChildren, useEffect } from 'react';

const AuthContext = createContext<{
  signIn: (session: {
    username: string;
    onboarding: boolean
  }) => void;
  finishOnboarding: (username: string) => void;
  signOut: () => void;
  session?: {
    username: string;
    onboarding: boolean
  } | null;
  isLoading: boolean;
}>({
  signIn: () => null,
  signOut: () => null,
  finishOnboarding: () => null,
  session: null,
  isLoading: false,
});

// This hook can be used to access the user info.
const useSession = () => {
  const value = useContext(AuthContext);

  if (process.env.NODE_ENV !== 'production') {
    if (!value) {
      throw new Error('useSession must be wrapped in a <SessionProvider />');
    }
  }

  return value;
}

const SessionProvider = ({ children }: PropsWithChildren) => {
  const [[isLoading, session], setSession] = useStorageState('session');

  useEffect(() => {
    // return () => setSession(null)
  }, [])

  return (
    <AuthContext.Provider
      value={{
        signIn: (session:{
          username: string;
          onboarding: boolean
        }) => {
          // Perform sign-in logic here
          setSession(session);
        },
        finishOnboarding: (username: string) => {
          setSession({
            username,
            onboarding: false
          })
        },
        signOut: () => {
          setSession(null);
        },
        session,
        isLoading,
      }}>
      {children}
    </AuthContext.Provider>
  );
}

export {
  useSession,
  SessionProvider
}
