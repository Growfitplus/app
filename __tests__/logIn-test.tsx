import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { router } from 'expo-router';
import { useSession } from '@/contexts/session';
import Login from '@/app/index';

jest.mock('expo-router', () => ({
  router: {
    push: jest.fn(),
  },
}));

jest.mock('@/contexts/session', () => ({
  useSession: jest.fn(),
}));

describe('Login', () => {
  const mockPush = jest.fn();
  const mockSignIn = jest.fn();

  beforeEach(() => {
    router.push = mockPush;
    (useSession as jest.Mock).mockReturnValue({
      session: null,
      signIn: mockSignIn,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(<Login />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const { getByText } = render(<Login />);

    expect(getByText('Ingresa con')).toBeTruthy();
    expect(getByText('Google')).toBeTruthy();
  });

  it('should call signIn when the button is pressed', () => {
    const { getByText } = render(<Login />);

    fireEvent.press(getByText('Google'));
    expect(mockSignIn).toHaveBeenCalledWith({
      onboarding: true,
      username: 'eamzea',
    });
  });

  it('should navigate to onboarding if session onboarding is true', () => {
    (useSession as jest.Mock).mockReturnValue({
      session: { onboarding: true },
      signIn: mockSignIn,
    });
    render(<Login />);
    expect(mockPush).toHaveBeenCalledWith('/(stack)/(onboarding)/start');
  });

  it('should navigate to home if session onboarding is false', () => {
    (useSession as jest.Mock).mockReturnValue({
      session: { onboarding: false },
      signIn: mockSignIn,
    });
    render(<Login />);
    expect(mockPush).toHaveBeenCalledWith('/(stack)/(tabs)/home');
  });
});
