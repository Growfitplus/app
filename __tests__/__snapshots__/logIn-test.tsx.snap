// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#00FFC2",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <RNSVGSvgView
      bbHeight={58}
      bbWidth={207}
      fill="none"
      focusable={false}
      height={58}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 58,
            "width": 207,
          },
        ]
      }
      width={207}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M160.236 13.678h-4.525c-.374 0-.698.133-.96.396a1.32 1.32 0 0 0-.396.963v25.431c0 .375.134.7.396.962.262.263.586.397.96.397h4.525c.412 0 .736-.134.959-.397.263-.263.396-.587.396-.962V15.037c0-.377-.133-.7-.393-.96-.226-.265-.551-.4-.962-.4ZM182.014 35.721h-3.761c-1.286 0-2.208-.416-2.741-1.234-.502-.826-.756-1.995-.756-3.473V19.566h6.768c.374 0 .697-.133.959-.396.263-.263.396-.588.396-.963v-3.17c0-.375-.133-.699-.396-.962a1.31 1.31 0 0 0-.959-.397h-6.768v-9.39c0-.375-.133-.699-.396-.962a1.31 1.31 0 0 0-.96-.396h-4.416c-.411 0-.753.133-1.017.399a1.432 1.432 0 0 0-.338.96V31.56c0 2.255.368 4.175 1.095 5.704a7.412 7.412 0 0 0 3.347 3.41c1.49.765 3.35 1.153 5.528 1.153h4.416c.375 0 .698-.134.96-.397a1.32 1.32 0 0 0 .396-.962v-3.39c0-.374-.134-.698-.396-.961a1.316 1.316 0 0 0-.961-.397ZM150.78 1.29h-5.524c-2.395 0-4.4.369-5.963 1.096-1.564.73-2.764 1.84-3.564 3.3-.801 1.42-1.206 3.249-1.206 5.434v29.348c0 .375.134.7.396.962.262.263.586.397.96.397h4.47c.374 0 .698-.134.96-.397a1.32 1.32 0 0 0 .396-.962V19.565h6.713c.413 0 .736-.133.96-.396a1.32 1.32 0 0 0 .396-.962v-3.17c0-.376-.134-.7-.394-.96-.226-.265-.55-.4-.963-.4h-6.713v-2.284c0-1.368.272-2.424.808-3.142.534-.713 1.53-1.074 2.961-1.074h5.307c.412 0 .736-.134.96-.397.262-.263.395-.587.395-.962V2.649c0-.375-.133-.7-.393-.96-.225-.265-.55-.4-.962-.4ZM34.682 40.565l-4.262-3.56a15.91 15.91 0 0 0 3.37-9.841c0-2.411-.522-4.729-1.551-6.891a3.146 3.146 0 0 0 2.258-3.587 3.131 3.131 0 0 0-3.637-2.524l-2.99.547a16.02 16.02 0 0 0-10.09-3.595c-8.83 0-16.014 7.203-16.014 16.054 0 8.853 7.184 16.054 16.014 16.054 2.692 0 5.355-.686 7.701-1.984l3.377 2.822a12.728 12.728 0 0 1-5.966 5.31 12.726 12.726 0 0 1-9.824.176 12.768 12.768 0 0 1-7.07-6.842 3.105 3.105 0 0 0-1.723-1.667 3.1 3.1 0 0 0-2.394.042A3.143 3.143 0 0 0 .26 45.207a19.087 19.087 0 0 0 4.248 6.137 18.946 18.946 0 0 0 6.27 4.038 18.971 18.971 0 0 0 6.987 1.33h.345a18.958 18.958 0 0 0 7.282-1.587 19.038 19.038 0 0 0 6.122-4.26 19.008 19.008 0 0 0 4.028-6.283c.096-.241.162-.5.196-.766a3.192 3.192 0 0 0-1.056-3.252Zm-7.148-13.397c0 5.393-4.376 9.78-9.755 9.78-5.38 0-9.756-4.387-9.756-9.78 0-5.392 4.377-9.78 9.756-9.78 5.38 0 9.756 4.388 9.756 9.78Z"
          fill={
            {
              "payload": 4280494648,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
        <RNSVGPath
          d="M203.219 14.537v-.002a1.257 1.257 0 0 0-.912-.366h-4.499V9.632c0-.367-.14-.687-.406-.927a1.225 1.225 0 0 0-.87-.352h-2.345c-.358 0-.673.126-.912.366-.239.239-.365.555-.365.914v4.537h-4.526c-.345 0-.659.118-.914.341l-.027.026c-.238.24-.365.556-.365.915v2.092c0 .36.127.675.365.915l.027.026c.254.222.57.34.914.34h4.526v4.68c0 .36.125.676.365.915.238.24.554.366.912.366h2.345c.333 0 .634-.122.87-.353.267-.24.406-.561.406-.927v-4.681h4.499c.358 0 .673-.127.912-.366.239-.24.365-.556.365-.915v-2.092a1.266 1.266 0 0 0-.365-.915Z"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
        <RNSVGPath
          d="M56.458 13.426h-2.235c-1.744 0-3.31.314-4.655.934a8.291 8.291 0 0 0-3.148 2.576V14.84a1.3 1.3 0 0 0-.45-1.018 1.312 1.312 0 0 0-.96-.396h-4.416c-.374 0-.698.133-.96.396s-.396.606-.396 1.018v25.741c0 .375.134.7.396.962.262.263.586.397.96.397h4.403c.41 0 .752-.134 1.016-.397.262-.263.395-.587.395-.962v-15.03c0-1.692.812-3.04 1.742-4.01.968-.97 2.333-1.462 4.056-1.462h4.252c.413 0 .736-.135.96-.397.262-.263.395-.587.395-.962v-3.935c0-.376-.133-.7-.393-.96-.226-.264-.55-.4-.962-.4ZM131.375 15.14c.062-.17.096-.325.096-.464 0-.341-.134-.629-.393-.851a1.079 1.079 0 0 0-.854-.399h-4.034c-.446 0-.807.134-1.069.397-.224.224-.357.454-.394.675l-5.299 17.942-1.136-3.592-4.487-14.186a1.462 1.462 0 0 0-.182-.394l-.003-.006a2.346 2.346 0 0 0-.317-.386c-.228-.26-.561-.41-.989-.444a2.542 2.542 0 0 0-.19-.007h-2.726c-.067 0-.131.002-.194.008-.427.034-.758.184-.985.443-.033.033-.063.066-.094.099-.087.095-.16.191-.223.287l-.003.006a1.749 1.749 0 0 0-.182.392l-4.596 14.531-1.027 3.247-5.3-17.942c-.036-.222-.169-.451-.393-.676-.262-.262-.622-.396-1.07-.396h-4.033c-.34 0-.628.134-.854.399-.259.222-.393.51-.393.85 0 .15.019.317.056.506l7.799 25.365c.062.136.129.264.199.384.115.237.27.442.465.613l.002.003c.26.26.657.394 1.176.394H103.29c.518 0 .933-.133 1.23-.395.296-.259.503-.598.614-1.005l5.624-17.133 5.625 17.133c.111.407.318.746.614 1.006.298.261.712.394 1.23.394H121.771c.519 0 .916-.134 1.176-.395l.002-.002a1.97 1.97 0 0 0 .465-.613c.07-.12.136-.248.199-.384l7.798-25.365c.005-.008-.009-.022-.036-.04ZM74 11.185c-8.818 0-15.992 7.192-15.992 16.032S65.182 43.249 74 43.249c8.818 0 15.993-7.192 15.993-16.032S82.819 11.185 74 11.185Zm0 25.424c-5.167 0-9.368-4.211-9.368-9.392 0-5.18 4.2-9.392 9.369-9.392 5.168 0 9.369 4.211 9.369 9.392 0 5.18-4.201 9.392-9.37 9.392Z"
          fill={
            {
              "payload": 4280494648,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    style={
      {
        "gap": 20,
      }
    }
  >
    <Text
      style={
        {
          "fontFamily": "RobotoMedium",
          "fontSize": 14,
          "textAlign": "center",
        }
      }
    >
      Ingresa con
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 16,
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "space-between",
          "paddingHorizontal": 16,
          "width": 312,
        }
      }
    >
      <Text />
      <Text
        style={
          {
            "fontFamily": "RobotoBold",
            "fontSize": 16,
          }
        }
      >
        Google
      </Text>
      <Text />
    </View>
  </View>
</RCTSafeAreaView>
`;
